src/soc/ibm/power9/istep_14_3.c

```diff
@@ -2,37 +2,501 @@
 
 #include <cpu/power/istep_14.h>
 
-static void p9_pcie_config(void)
+#include <console/console.h>
+#include <cpu/power/scom.h>
+
+#include "pci.h"
+
+static uint64_t pec_addr(uint8_t pec, uint64_t addr)
+{
+	return addr + pec*0x400;
+}
+
+static void init_pecs(const uint8_t *iovalid_enable)
+{
+	enum {
+		P9N2_PEC_ADDREXTMASK_REG = 0x4010C05,
+		PEC_PBCQHWCFG_REG = 0x4010C00,
+		PEC_NESTTRC_REG = 0x4010C03,
+		PEC_PBAIBHWCFG_REG = 0xD010800,
+
+		/* powerbus.c has these too */
+		MBOX_SCRATCH_REG1 = 0x00050038,
+		MBOX_SCRATCH_REG6_GROUP_PUMP_MODE = (1 << 23),
+
+		PEC_PBCQHWCFG_REG_PE_DISABLE_TCE_ARBITRATION = 60,
+		PEC_PBAIBHWCFG_REG_PE_PCIE_CLK_TRACE_EN = 30,
+		PEC_AIB_HWCFG_OSBM_HOL_BLK_CNT = 42,
+		PEC_PBCQHWCFG_REG_PE_DISABLE_OOO_MODE = 0x16,
+		PEC_PBCQHWCFG_REG_PE_DISABLE_WR_SCOPE_GROUP = 42,
+		PEC_PBCQHWCFG_REG_PE_CHANNEL_STREAMING_EN = 33,
+		PEC_PBCQHWCFG_REG_PE_DISABLE_WR_VG = 41,
+		PEC_PBCQHWCFG_REG_PE_DISABLE_INTWR_VG = 43,
+		PEC_PBCQHWCFG_REG_PE_DISABLE_INTWR_SCOPE_GROUP = 44,
+		PEC_PBCQHWCFG_REG_PE_DISABLE_RD_SCOPE_GROUP = 51,
+		PEC_PBCQHWCFG_REG_PE_DISABLE_RD_VG = 54,
+		PEC_PBCQHWCFG_REG_PE_DISABLE_TCE_SCOPE_GROUP = 56,
+		PEC_PBCQHWCFG_REG_PE_DISABLE_TCE_VG = 59,
+	};
+
+	uint64_t scratch_reg6 = 0;
+	uint8_t pec = 0;
+	bool node_pump_mode = false;
+	uint8_t dd = get_dd();
+
+	scratch_reg6 = read_scom(MBOX_SCRATCH_REG1 + 5);
+
+	/* ATTR_PROC_FABRIC_PUMP_MODE, it's either node or group pump mode */
+	node_pump_mode = !(scratch_reg6 & MBOX_SCRATCH_REG6_GROUP_PUMP_MODE);
+
+	for (pec = 0; pec < MAX_PEC_PER_PROC; ++pec) {
+		uint64_t val = 0;
+
+		printk(BIOS_EMERG, "Initializing PEC%d...\n", pec);
+
+		/*
+		 * ATTR_FABRIC_ADDR_EXTENSION_GROUP_ID = 0
+		 * ATTR_FABRIC_ADDR_EXTENSION_CHIP_ID = 0
+		 */
+		scom_and_or_for_chiplet(N2_CHIPLET_ID, pec_addr(pec, P9N2_PEC_ADDREXTMASK_REG),
+					~PPC_BITMASK(0, 6),
+					PPC_SHIFT(0, 6));
+
+		/*
+		 * Phase2 init step 1
+		 * NestBase + 0x00
+		 * Set bits 00:03 = 0b0001 Set hang poll scale
+		 * Set bits 04:07 = 0b0001 Set data scale
+		 * Set bits 08:11 = 0b0001 Set hang pe scale
+		 * Set bit 22 = 0b1 Disable out­of­order store behavior
```

Comment:
Lint complains about non-ASCII symbols in this line, probably nonbreakable spaces in `out of order` as they are visible in vi.